Newsletters: add 123 (2020-11-11) (#483)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-11-11-newsletter.md

      smaller than 10^2, so to calculate the remainder we can add the bottom half
      (37 in this case) to the top half times the difference of the modulus from the
      base number power (14 * (10^2 - 99) in this case). The same technique is used
      in the MuHash implementation and explains why the modulus was chosen as a
      number slightly smaller than 2^3072."
  a3link="https://bitcoincore.reviews/19055-2#l-246"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

FIXME:none as of Sunday; harding to update or delete this section Tuesday

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Hardware Wallet Interface (HWI)][hwi repo], [Bitcoin Improvement Proposals
(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [BOLTs #807][] Fail channel in case of high-S remote signature reception FIXME:dongcarl

> - [Eclair #1545][] adds a blockchain watchdog to detect eclipse attacks. Eclair

Reviewer: perhaps link "eclipse attacks" to the optech topic page

Reviewer 2: Good idea. Done!

  fetches block headers from multiple third-party sources to countercheck the
  best chain provided by its peers. As described in [Newsletter #101][news101 time-dilation],
  an attacker could use an eclipse attack to hide a malicious channel closing
  attempt until the dispute period expires.

> - [LND #4735][] adds as `max_local_csv` configuration parameter that

Reviewer: s/as/a/

maybe s/parameter/option/ like the writeup on LND #4701 in the next paragraph (and Bitcoin Core)

  will reject channels from peers who require the local node wait more
  than the indicated number of blocks before being able to spend its
  funds if it unilaterally closes the channel.  The default maximum is

>   1,000 blocks, and there's also a minimum of 144 blocks to ensure users

Reviewer: s/1,000/10,000/? from looking at https://github.com/lightningnetwork/lnd/pull/4735/files#diff-0e426a43248661127a0c0ee115aef7a1093b635f8993b3f7ebb1dd9f05b8f249R119
```go
	// defaultMaxLocalCSVDelay is the maximum delay we accept on our
	// commitment output.
	// TODO(halseth): find a more scientific choice of value.
	defaultMaxLocalCSVDelay = 10000
```

  don't set the value lower than is reasonable.

- [LND #4701][] enables the `assumechanvalid` (assume channels are

>   valid) configuration option by default.  For nodes using the Neutrino

Reviewer: s/using/that use/ might be clearer

  client to retrieve [compact block filters][topic compact block

>   filters], this will assume the channels the node learns about via LN

Reviewer: maybe s/this will assume/assumechanvalid assumes/

  gossip are available.  If they're not actually available (e.g. they've
  been closed recently), any payments LND attempts to route through
  those channels will fail.  This may cause delays or unnecessary

>   payment failures but not the loss of funds.

Reviewer: Perhaps mention the tradeoffs a bit, e.g. from this paragraph in the pull, the option avoids long initial verification rescans on startup, but at the risk of an incorrect view of the network.

```
; Skip checking channel spentness and existence during graph validation for
; neutrino. Enabling this option means that neutrino nodes will not need to
; perform long rescans which block initial usage of the daemon, but comes at
; the cost of not validating channels in your routing graph. Skipping this
; validation means that your node may have an incorrect view of the network
; if it receives updates for closed or non-existent channels. This could affect
; routing, but funds are safu.
```

Reviewer: ￼Discussed with PR author Carla, who very helpfully mentioned, "Just a nit on assumechanvalid, it’s not enabled by default, we just removed the experimental build flag (people still need to opt in)." The reason for the PR: "people were confused when we told them about it but it wasn’t included in the binary when they built it."

Author: (1) Thanks for noticing that I focused on the downside but forgot to describe the upside; (2) HUGE thanks for verifying with Carla and allowing me to fix my error before we published.  I appreciate it!

Reviewer: Thanks to Carla for having a look! 🌳

      added items) and a denominator (for the removed items), which would require
      6144 bits (768 bytes) of internal state, but would reduce the number of costly
      modulo operations required."
  a0link="https://bitcoincore.reviews/19055-2#l-47"

  q1="How can we test for membership in the MuHash set?"
  a1="MuHash is not an accumulator, so there is no efficient way to test for set
      membership. The only way would be to recalculate the MuHash digest from the
      entire set of objects and then compare the digests."
  a1link="https://bitcoincore.reviews/19055-2#l-131"

  q2="What does the `#ifdef HAVE___INT128` code in the MuHash implementation do?"
  a2="`#ifdef HAVE___INT128` is a C++ preprocessor directive. It ensures that code
      contained within the `#ifdef` and `#endif` directives is only compiled on
      platforms that support 128-bit integers. Since the MuHash code is
      operating on numbers that are many times larger than built-in integer types
      can handle, those numbers are broken down into multiple 'limbs', which are
      combined to make the full number. Larger limbs means fewer operations and

>       therefore [better performance][muhash performance]. Support for 128 bit

Reviewer: one last hyphen managed to get away (s/128 bit/128-bit/)

Reviewer 3: Thanks! Fixed.

---
title: 'Bitcoin Optech Newsletter #123'
permalink: /en/newsletters/2020/11/11/
name: 2020-11-11-newsletter
slug: 2020-11-11-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter shares the announcement of a marketplace for
incoming LN channels Also included are our regular sections with summaries of a
Bitcoin Core PR Review Club meeting and notable changes to popular Bitcoin
infrastructure software.

## Action items

*None this week.*

## News

- **Incoming channel marketplace:** Olaoluwa Osuntokun [announced][pool
  announce] a new marketplace for buying incoming LN channels.  Users and
  merchants need channels with incoming capacity in order to be able to

>   quickly receive funds over LN.  Some existing node operators already

Reviewer 4: split infinitive

```suggestion
  receive funds over LN quickly.  Some existing node operators already
```

  provide incoming channels, either for free or as a paid service, but
  the Lightning Pool marketplace hopes to make this service more
  standardized and competitive.  The initial focus is a contract for

>   [highly ranked][node rankings] node operators to provide an incoming

Reviewer 4: (feel free to ignore) I think the node rankings link is cool, but it's a little hard to understand what it is given that it's not sorted.

      base number power (14 * (10^2 - 99) in this case). The same technique is used
      in the MuHash implementation and explains why the modulus was chosen as a
      number slightly smaller than 2^3072."
  a3link="https://bitcoincore.reviews/19055-2#l-246"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

FIXME:none as of Sunday; harding to update or delete this section Tuesday

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Hardware Wallet Interface (HWI)][hwi repo], [Bitcoin Improvement Proposals
(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [BOLTs #807][] amends [BOLT2][] to specify that a node should treat the
  receipt of non-standard high-*S* signatures the same as invalid signatures and
  immediately fail the channel or connection. This mitigates against

>   [CVE-2020-26895][] covered previously in [Newsletter #121][news121

Reviewer 4: CVE-2020-26895 isn't linked

Reviewer 5: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-26895

  lows-sig-cve].

- [Eclair #1545][] adds a blockchain watchdog to detect eclipse attacks. Eclair
  fetches block headers from multiple third-party sources to countercheck the
  best chain provided by its peers. As described in [Newsletter #101][news101 time-dilation],
  an attacker could use an eclipse attack to hide a malicious channel closing
  attempt until the dispute period expires.

- [LND #4735][] adds as `max_local_csv` configuration parameter that

>   will reject channels from peers who require the local node wait more
>   than the indicated number of blocks before being able to spend its

Reviewer 4: nit: a little wordy

```suggestion
  will reject channels from peers who require the local node to wait
  for more than the indicated number of blocks before spending its
```

  the Lightning Pool marketplace hopes to make this service more
  standardized and competitive.  The initial focus is a contract for
  [highly ranked][node rankings] node operators to provide an incoming
  channel for a period of 2,016 blocks (about two weeks).  Additional
  contract lengths and other features are planned for the future.

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting.*

[Add MuHash3072 implementation][review club #19055-2] is a PR ([#19055][Bitcoin
Core #19055]) by Fabian Jahr (incorporating code originally written by Pieter
Wuille) that implements the [MuHash algorithm][muhash mailing list] in C++.
MuHash is a rolling hash algorithm that can be used to calculate the hash
digest of a set of objects and efficiently update that digest when items are
added to or removed from the set. The algorithm could be used to calculate a
digest of the complete UTXO set, which would be useful for database consistency

> checks or for a fast sync method such as [AssumeUTXO][topic assumeutxo].

Reviewer 5: Recommending s/AssumeUTXO/assumeUTXO

Reviewer 3: We seem to be inconsistent here, using both assumeutxo (https://bitcoinops.org/en/newsletters/2019/06/12/#assume-utxo-demo) and AssumeUTXO (https://bitcoinops.org/en/newsletters/2019/12/28/#assumeutxo). Perhaps one to add to the style guide?

For now, I've changed this to assumeutxo.

The high-level concepts for MuHash had been discussed in [a previous review
club meeting][review club #19055], so this meeting focused specifically on the
specification and implementation of the algorithm.

{% include functions/details-list.md
  q0="How much state is stored inside the MuHash3072 rolling hash object? How
      much data is returned to a user requesting the set hash?"
  a0="The MuHash3072 object stores 3072 bits (384 bytes) of internal state.
      This state is hashed using a 256-bit hash function to return 256 bits (32
      bytes) to the end user. An optimization is to store both a numerator (for the
      added items) and a denominator (for the removed items), which would require
      6144 bits (768 bytes) of internal state, but would reduce the number of costly

>       modulus operations required."

Reviewer 5: Im above my pay grade here, but s/modulus/modulo ?

Reviewer 3: Absolutely correct.

- [BOLTs #807][] Fail channel in case of high-S remote signature reception FIXME:dongcarl

- [Eclair #1545][] adds a blockchain watchdog to detect eclipse attacks. Eclair
  fetches block headers from multiple third-party sources to countercheck the
  best chain provided by its peers. As described in [Newsletter #101][], an
  attacker could use an eclipse attack to hide a malicious channel closing

>   attempt until the dispute period expires.

Reviewer 2: First PR description I did. :)

I've been considering to switch the order of sentence two and three, i.e. first explain the attack, then the defense. Any preferences on that?

Reviewer: Seems good to me as-is :)

  an attacker could use an eclipse attack to hide a malicious channel closing
  attempt until the dispute period expires.

- [LND #4735][] adds a `max_local_csv` configuration parameter that
  will reject channels from peers who require the local node to wait for more
  than the indicated number of blocks before spending its
  funds if it unilaterally closes the channel.  The default maximum is
  10,000 blocks, and there's also a minimum of 144 blocks to ensure users
  don't set the value lower than is reasonable.

- [LND #4701][] adds the `assumechanvalid` (assume channels are
  valid) configuration option to default builds.  For nodes using the Neutrino
  client to retrieve [compact block filters][topic compact block
  filters], this allows the node to assume the channels it learns about via LN

>   gossip are available rather than by expending additional bandwidth

Reviewer: omit "by"?

Author: I think it's better with it.
